helpers: more helpers

Fang- opened the pull request:
9725793 does a light reorganization. I think splitting this up will be more sustainable as these sub-suites grow.

13e8fb6 provides nicer affordances for doing typechecking for `Noun` objects.  
I tried briefly, but in hindsight it's obvious we can't do any `isShape(myNoun, ['a', 'c', 'n'])` (for `?=([@ ^ *] my-noun)`) style shenanigans. So I think the `isMap` style checkers are the best we can do. Maybe if you get a sufficient amount of those into the stdlib you're covering most of your bases, anyway.  
We should do a pass over the codebase to replace all `instanceof` code with these helpers, but that's out of scope for this PR.

06e2549 provides `Noun`-level set/map creation helpers. 2a9316f wraps those to provide "from non-`Noun` data" versions for the `dejs` part of the suite.

c43bc86 and 89a0fc2 are attempts at simplifying some of the logic we wrote for the verifier stuff earlier this week.

There's probably stuff that I missed, may tack more on later as I write helpers for verifier work, but this should be a good start.

Fang- committed:
helpers: split into enjs and dejs files

Fang- committed:
std helpers: basic set and map creation helpers

Fang- committed:
noun: provide typechecking helpers

Should use these more broadly, but that would explode the linecount of
the PR I want to make with largely unrelated changes...

Fang- committed:
dejs helpers: map and set creation

Fang- committed:
enjs helpers: frond w/o wrapper object

Fang- committed:
enjs helpers: tuple for "flat" cell reading

Fang- committed:
enjs helpers: less stupid tuple function

This one can at least process the full noun always.

Fang- committed:
enjs helpers: string from tape

arthyn commented on src/noun-enjs.ts:
```diff
@@ -0,0 +1,204 @@
+  };
+};
+
+const array = function (item: EnjsFunction): (n: Noun) => Json[] {
+  return function (noun) {
+    let a: Json[] = [];
+    while (noun instanceof Cell) {
+      a.push(item(noun.head));
+      noun = noun.tail;
+    }
+    return a;
+  };
+};
```
will this grab the last noun?

Fang- replied:
This goes from `(list *)` to `any[]`. I put a comment clarifying as such. So, in some sense, no, we intentionally ignore the list's trailing `~`.

arthyn commented on src/noun-enjs.ts:
```diff
@@ -0,0 +1,204 @@
+  return function (noun) {
+    let a: Json[] = [];
+    if (noun instanceof Cell) {
+      if (!(noun.tail instanceof Cell)) {
```
is this guaranteed to be the case at every level of the tree?

Fang- replied:
Yes, see also the definition for `+tree` in hoon.hoon. Every node is either `~` or a triple (with `n`ode value, `l`eft & `r`ight branches.)

arthyn reviewed (approved):
lgtm just two questions

Fang- committed:
tests: add enjs tests

Fang- committed:
enjs: don't mangle argument array in tuple()

Calling .splice() on the argument array modifies it in-place, which
might affect the callsite. Iterate over it by keeping an index pointer
instead.

Fang- committed:
enjs: small touch-ups

Cleaner error message, comments for clarity.
